torchao/prototype/sparsity/PartialLinear/partial_linear.py

```diff
@@ -0,0 +1,183 @@
+    input features by weight magnitude: :math:`y = x * (M \odot W^T) + b`,
+    where $\odot$ is the element-wise product and $M$ is a binary mask.
+    This module implements a form of structured sparsity that can reduce computation
+    and memory usage during inference.
+    Args:
+        in_features: size of each input sample
+        out_features: size of each output sample
+        top_k: number of weights to retain per output feature (default: in_features // 2)
+        bias: If set to ``False``, the layer will not learn an additive bias.
+            Default: ``True``
+        update_mask_every: update the mask every N forward passes during training (default: 50)
+    Shape:
+        - Input: :math:`(*, H_\text{in})` where :math:`*` means any number of
+          dimensions including none and :math:`H_\text{in} = \text{in\_features}`.
+        - Output: :math:`(*, H_\text{out})` where all but the last dimension
+          are the same shape as the input and :math:`H_\text{out} = \text{out\_features}`.
+    Attributes:
+        weight: the learnable weights of the module of shape
+            :math:`(\text{out\_features}, \text{in\_features})`. The values are
+            initialized from :math:`\mathcal{U}(-\sqrt{k}, \sqrt{k})`, where
+            :math:`k = \frac{1}{\text{in\_features}}`
+        mask: binary mask of shape :math:`(\text{out\_features}, \text{in\_features})`
+            indicating which weights are retained (1) or pruned (0)
+        bias:   the learnable bias of the module of shape :math:`(\text{out\_features})`.
+                If :attr:`bias` is ``True``, the values are initialized from
+                :math:`\mathcal{U}(-\sqrt{k}, \sqrt{k})` where
+                :math:`k = \frac{1}{\text{in\_features}}`
+        is_sparse_forward: when True, uses a completely sparse computation for forward pass
+    Examples::
+        >>> m = nn.PartialLinear(20, 30, top_k=5)  # Each output connected to top 5 inputs
+        >>> input = torch.randn(128, 20)
+        >>> output = m(input)
+        >>> print(output.size())
+        torch.Size([128, 30])
+    """
+    __constants__ = ["in_features", "out_features", "top_k", "update_mask_every", "is_sparse_forward"]
+    in_features: int
+    out_features: int
+    top_k: int
+    update_mask_every: int
+    is_sparse_forward: bool
+    weight: Tensor
+
+    def __init__(
+        self,
+        in_features: int,
+        out_features: int,
+        top_k: int = None,
+        bias: bool = True,
+        update_mask_every: int = 50,
+        is_sparse_forward: bool = False,
+        device=None,
+        dtype=None,
+    ) -> None:
+        factory_kwargs = {"device": device, "dtype": dtype}
+        super().__init__()
+
+        self.in_features = in_features
+        self.out_features = out_features
+
+        # Default to half of the input features if not specified
+        if top_k is None:
```

Comment:
I think better to just throw a value error or make this an arg instead of a kwarg